File: shared/go_libs/crypto/hashing.go
package crypto

import (
	"fmt"

	"golang.org/x/crypto/bcrypt"
)

const (
	// DefaultBcryptCost is the default cost for bcrypt hashing
	DefaultBcryptCost = 12
)

// HashPassword hashes a password using bcrypt with the default cost.
func HashPassword(password string) (string, error) {
	return HashPasswordWithCost(password, DefaultBcryptCost)
}

// HashPasswordWithCost hashes a password using bcrypt with a specified cost.
// Cost should be between 4 and 31 (10-12 recommended for production).
func HashPasswordWithCost(password string, cost int) (string, error) {
	if password == "" {
		return "", fmt.Errorf("password cannot be empty")
	}

	if cost < bcrypt.MinCost || cost > bcrypt.MaxCost {
		return "", fmt.Errorf("invalid bcrypt cost: must be between %d and %d", bcrypt.MinCost, bcrypt.MaxCost)
	}

	hashedBytes, err := bcrypt.GenerateFromPassword([]byte(password), cost)
	if err != nil {
		return "", fmt.Errorf("failed to hash password: %w", err)
	}

	return string(hashedBytes), nil
}

// VerifyPassword verifies that a plaintext password matches a bcrypt hash.
// Returns true if the password is correct, false otherwise.
func VerifyPassword(password, hash string) (bool, error) {
	if password == "" || hash == "" {
		return false, fmt.Errorf("password and hash cannot be empty")
	}

	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
	if err != nil {
		if err == bcrypt.ErrMismatchedHashAndPassword {
			return false, nil
		}
		return false, fmt.Errorf("failed to verify password: %w", err)
	}

	return true, nil
}

File: shared/node_libs/src/crypto/hashing.ts
import * as argon2 from 'argon2';

/**
 * Default options for Argon2 hashing.
 * These provide a good balance of security and performance for most applications.
 */
const DEFAULT_OPTIONS: argon2.Options = {
  type: argon2.argon2id, // Hybrid mode (resistant to both GPU and side-channel attacks)
  memoryCost: 65536,     // 64 MB
  timeCost: 3,           // 3 iterations
  parallelism: 4,        // 4 parallel threads
};

/**
 * Hashes a password using Argon2id with default secure settings.
 */
export async function hashPassword(password: string): Promise<string> {
  if (!password) {
    throw new Error('Password cannot be empty');
  }

  return argon2.hash(password, DEFAULT_OPTIONS);
}

/**
 * Hashes a password using Argon2id with custom options.
 */
export async function hashPasswordWithOptions(
  password: string,
  options: Partial<argon2.Options> = {}
): Promise<string> {
  if (!password) {
    throw new Error('Password cannot be empty');
  }

  return argon2.hash(password, { ...DEFAULT_OPTIONS, ...options });
}

/**
 * Verifies that a plaintext password matches an Argon2 hash.
 * Returns true if the password is correct, false otherwise.
 */
export async function verifyPassword(
  password: string,
  hash: string
): Promise<boolean> {
  if (!password || !hash) {
    throw new Error('Password and hash cannot be empty');
  }

  try {
    return await argon2.verify(hash, password);
  } catch (error) {
    // argon2.verify throws on mismatch in some versions
    if (error instanceof Error && error.message.includes('verification failed')) {
      return false;
    }
    throw error;
  }
}
